### PR TITLE
Trying to improve QUIC reassembler (#1195)

### PR DESCRIPTION
**Pull request by viniciussn**
This PR changes QUIC reassembler to accept out-of-order fragments.
Almost all testes passed, only  "quic_frags_ch_out_of_order_same_packet_craziness.pcapng" has different output:
<details>
  <summary>Old one</summary>
  
```
-----------------------------------------------------------
* NOTE: This is demo app to show *some* nDPI features.
* In this demo we have implemented only some basic features
* just to show you what you can do with the library. Feel 
* free to extend it and send us the patches for inclusion
------------------------------------------------------------

Using nDPI (4.3.0-3530-b83e9f20) [1 thread(s)]
Using libgcrypt version 1.8.6internal
Reading packets from pcap file tests/pcap/quic_frags_ch_out_of_order_same_packet_craziness.pcapng...
Running thread 0...

nDPI Memory statistics:
        nDPI Memory (once):      237.23 KB    
        Flow Memory (per flow):  688 B        
        Actual Memory:           7.72 MB      
        Peak Memory:             7.72 MB      
        Setup Time:              26 msec
        Packet Processing Time:  5 msec

Traffic statistics:
        Ethernet bytes:        253464        (includes ethernet CRC/IFC/trailer)
        Discarded bytes:       0            
        IP packets:            179           of 179 packets total
        IP bytes:              249168        (avg pkt size 1392 bytes)
        Unique flows:          113          
        TCP Packets:           0            
        UDP Packets:           179          
        VLAN Packets:          0            
        MPLS Packets:          0            
        PPPoE Packets:         0            
        Fragmented Packets:    0            
        Max Packet size:       1358         
        Packet Len < 64:       0            
        Packet Len 64-128:     0            
        Packet Len 128-256:    0            
        Packet Len 256-1024:   0            
        Packet Len 1024-1500:  179          
        Packet Len > 1500:     0            
        nDPI throughput:       33.99 K pps / 367.22 Mb/sec
        Analysis begin:        19/May/2021 06:38:31
        Analysis end:          20/May/2021 11:32:22
        Traffic throughput:    0.00 pps / 19 b/sec
        Traffic duration:      104031.414 sec
        Guessed flow protos:   113          
        DPI Packets (UDP):     179           (1.58 pkts/flow)
        Confidence: DPI        113           (flows)


Detected protocols:
        Google               packets: 7             bytes: 9744          flows: 3            
        QUIC                 packets: 126           bytes: 175392        flows: 96           
        GoogleServices       packets: 4             bytes: 5568          flows: 1            
        Azure                packets: 39            bytes: 54288         flows: 10           
        GoogleCloud          packets: 3             bytes: 4176          flows: 3            


Protocol statistics:
        Acceptable                  249168 bytes
```

</details>

<details>
  <summary>New one</summary>
  
```
-----------------------------------------------------------
* NOTE: This is demo app to show *some* nDPI features.
* In this demo we have implemented only some basic features
* just to show you what you can do with the library. Feel 
* free to extend it and send us the patches for inclusion
------------------------------------------------------------

Using nDPI (4.3.0-3530-b83e9f20) [1 thread(s)]
Using libgcrypt version 1.8.6internal
Reading packets from pcap file tests/pcap/quic_frags_ch_out_of_order_same_packet_craziness.pcapng...
Running thread 0...

nDPI Memory statistics:
        nDPI Memory (once):      237.23 KB    
        Flow Memory (per flow):  688 B        
        Actual Memory:           7.08 MB      
        Peak Memory:             7.08 MB      
        Setup Time:              25 msec
        Packet Processing Time:  4 msec

Traffic statistics:
        Ethernet bytes:        253464        (includes ethernet CRC/IFC/trailer)
        Discarded bytes:       0            
        IP packets:            179           of 179 packets total
        IP bytes:              249168        (avg pkt size 1392 bytes)
        Unique flows:          113          
        TCP Packets:           0            
        UDP Packets:           179          
        VLAN Packets:          0            
        MPLS Packets:          0            
        PPPoE Packets:         0            
        Fragmented Packets:    0            
        Max Packet size:       1358         
        Packet Len < 64:       0            
        Packet Len 64-128:     0            
        Packet Len 128-256:    0            
        Packet Len 256-1024:   0            
        Packet Len 1024-1500:  179          
        Packet Len > 1500:     0            
        nDPI throughput:       36.72 K pps / 396.67 Mb/sec
        Analysis begin:        19/May/2021 06:38:31
        Analysis end:          20/May/2021 11:32:22
        Traffic throughput:    0.00 pps / 19 b/sec
        Traffic duration:      104031.414 sec
        Guessed flow protos:   0            
        DPI Packets (UDP):     113           (1.00 pkts/flow)
        Confidence: DPI        113           (flows)


Detected protocols:
        DataSaver            packets: 1             bytes: 1392          flows: 1            
        YouTube              packets: 16            bytes: 22272         flows: 16           
        Google               packets: 59            bytes: 82128         flows: 43           
        QUIC                 packets: 30            bytes: 41760         flows: 25           
        DoH_DoT              packets: 4             bytes: 5568          flows: 4            
        PlayStore            packets: 1             bytes: 1392          flows: 1            
        GoogleServices       packets: 48            bytes: 66816         flows: 17           
        Azure                packets: 19            bytes: 26448         flows: 5            
        GoogleCloud          packets: 1             bytes: 1392          flows: 1            


Protocol statistics:
        Safe                          1392 bytes
        Acceptable                  218544 bytes
        Fun                          29232 bytes

Risk stats [found 33 (29.2 %) flows with risks]:
        HTTP Suspicious User-Agent                   4 [ 6.2 %]
        TLS (probably) Not Carrying HTTPS           14 [21.9 %]
        Suspicious DGA Domain name                   3 [ 4.7 %]
        Missing SNI TLS Extension                    5 [ 7.8 %]
        TLS Suspicious Extension                    30 [46.9 %]
        Text Contains Non-Printable Characters       8 [12.5 %]

        NOTE: as one flow can have multiple risks set, the sum of the
              last column can exceed the number of flows with risks.
```

</details>

I believe there are other cases to deal with, but it's a start.

### CODE REVIEW

**Comment by utoni** [comment]
You need to commit any changes in `./tests/result/` as well.

In your case `git rm ./tests/result/quic_frags_ch_out_of_order_same_packet_craziness.pcapng.out` and `./tests/do.sh` should do the trick.

File: tests/result/quic_frags_ch_out_of_order_same_packet_craziness.pcapng.out
Comment (utoni) [inline tests/result/quic_frags_ch_out_of_order_same_packet_craziness.pcapng.out]: This line does not look good.

File: tests/result/quic_frags_ch_out_of_order_same_packet_craziness.pcapng.out
Comment (utoni) [inline tests/result/quic_frags_ch_out_of_order_same_packet_craziness.pcapng.out]: same here

File: tests/result/quic_frags_ch_out_of_order_same_packet_craziness.pcapng.out
Comment (utoni) [inline tests/result/quic_frags_ch_out_of_order_same_packet_craziness.pcapng.out]: and here

File: src/lib/protocols/quic.c
Comment (utoni) [inline src/lib/protocols/quic.c]: I do not fully understand what your intention was and how that should fix out-of-order packets.

**Review by utoni — APPROVED** [review]
@IvanNardi may give you a more detailed answer on what is going wrong here.

**Review by utoni — CHANGES_REQUESTED** [review]
Sorry, accidentally approved.

File: src/lib/protocols/quic.c
Comment (viniciussn) [reply src/lib/protocols/quic.c]: Previously, there was an if statement checking if we were getting the frames in order. 

```
if(flow->l4.udp.quic_reasm_buf_len != frag_offset)
    return -2; /* Out-of-order, retransmission, overlapping */
```

I removed that statement and started saving the frames in the buffer according to its offset.
I also modified (https://github.com/ntop/nDPI/pull/1489/commits/f273c5d4476d8abc5437e82374f71dc6c62d0de0) the "is_ch_complete" function. Now we need to check the buffer length and the last written position: if they don't match, frames are missing somewhere in the buffer.

Comment (utoni) [reply src/lib/protocols/quic.c]: LGTM. Waiting for @IvanNardi's approval. =)

File: src/lib/protocols/quic.c
Comment (IvanNardi) [inline src/lib/protocols/quic.c]: Could you update this comment, please?
Working: in-order and out-of order, non overlapping, fragments
Not supported: retransmissions and (partial) overlapping

File: src/lib/protocols/quic.c
Comment (IvanNardi) [inline src/lib/protocols/quic.c]: You need to move this new check in the previous "if" above; otherwise we end up reading un-initialized buffer (as you can easily check with valgrind)

**Review by IvanNardi — CHANGES_REQUESTED** [review]
This change adds support for the common case of out-of-order, non overlapping, non duplicated fragments. It doesn't handle overlapping or duplicate fragments.
This is fine and it is a great improvement to the current code.

However I am not yet sure that if we found overlapping or duplicate fragments we don't have any invalid memory access.
Let me run some tests: I'll update you soon

File: src/lib/protocols/quic.c
Comment (viniciussn) [reply src/lib/protocols/quic.c]: done!

File: src/lib/protocols/quic.c
Comment (viniciussn) [reply src/lib/protocols/quic.c]: done!

**Comment by viniciussn** [comment]
Do we have some pcaps with overlapping and duplicate fragments?

**Review by IvanNardi — APPROVED** [review]
I have not been able to capture/create/fuzz any traces triggering an error.
However I am not still convinced that we access `flow->l4.udp.quic_reasm_buf` safely in every scenarios.
Nonetheless this PR fixes a real issue and it is a visible improvement  compare to the current code.
Let's merge it and we will fix any issues if/when they will arise (if any)

**Comment by IvanNardi** [comment]
Thank you

**Comment by viniciussn** [comment]
Glad to help. I'm working on the other two cases and will create another PR as soon as possible.

**Pull request merged**
